bin/run.sh

```diff
@@ -1,55 +1,94 @@
-    )
+validate_input() {
+    if ! [[ -d "$in_dir" && -r "$in_dir" ]]; then
+        usage "Error: not a dir or not readable: ${in_dir@Q}"
+    fi
+    if ! [[ -d "$out_dir" && -w "$out_dir" ]]; then
+        usage "Error: not a dir or not writable: ${out_dir@Q}"
+    fi
+    local source="$in_dir/${snake_slug}.sh"
+    if ! [[ -f $source && -r $source ]]; then
+        usage "Error: not a file or not readable: ${source@Q}"
+    fi
 }
 
-process() {
-    # this reads stdin and emits to stdout
-    ruby bin/process-shellcheck-output.rb
+array_includes() {
+    local -n arr=$1
+    local elem=$2 e
+    for e in "${arr[@]}"; do [[ "$e" == "$elem" ]] && return; done
+    return 1
 }
 
-validate() {
-    if ! [[ -d "$inDir" && -r "$inDir" ]]; then
-        usage "Error: not a dir or not readable: ${inDir@Q}"
-    fi
-    if ! [[ -d "$outDir" && -w "$outDir" ]]; then
-        usage "Error: not a dir or not writable: ${outDir@Q}"
+merge_json() {
+    local files=("$out_dir"/!(expected)_analysis.json)
+    local summary n_comments temp
+
+    # Merge the comments in the json files into "analysis.json".
+    # Thanks to https://stackoverflow.com/a/36218044/7552
+    jq --slurp '{
+        comments: (reduce .[] as $item ([]; . + $item.comments))
+    }' "${files[@]}" > "$out_dir/analysis.json"
+
+    n_comments=$(jq '.comments | length' "$out_dir/analysis.json")
+
+    # Use the shellcheck summary if it exists
+    if array_includes files "$out_dir/shellcheck_analysis.json"; then
+        summary=$(jq -r .summary "$out_dir/shellcheck_analysis.json")
     fi
-    local source="$inDir/${filename}.sh"
-    if ! [[ -f $source && -r $source ]]; then
-        usage "Error: not a file or not readable: ${source@Q}"
+
+    # Otherwise, perhaps choose a different one
+    if [[ -z $summary || $summary == *"No Shellcheck suggestions"* ]]; then
+        if ((n_comments == 0)); then
+            summary="Congrats! No suggestions"
+        else
+            summary="Some comments"
+        fi
     fi
+
+    temp=$(mktemp)
+    jq --arg sum "$summary" '.summary = $sum' "$out_dir/analysis.json" > "$temp" \
+    && mv "$temp" "$out_dir/analysis.json"
 }
 
 main() {
     local slug=$1
-    local inDir=${2:-.}
-    local outDir=${3:-.}
-    local filename=${slug//-/_}
+    local in_dir=${2:-.}
+    local out_dir=${3:-.}
+    local snake_slug=${slug//-/_}
+
+    validate_input
+
+    # the analysis.bash is expected to have a namespaced `analyze` function.
+    # e.g.
+    #   "lib/foo-bar/analysis.bash"
+    # contains
+    #   foo_bar::analyze() { ... }
+    #
+    # Each analyze function will emit a *_analysis.json file
+    # containing that module's list of comments to show the student:
+    #   {"comments": [...]}
+
+    local namespace
+    for module in general shellcheck "$slug"; do
+        src_file="lib/$module/analysis.bash"
+        if [[ -f $src_file ]]; then
+            namespace=${module//-/_}
 
-    validate
+            source "$src_file" \
```

Comment:
sourcing random checks makes me a bit nervous. Would it make more sense to run them in a subshell?